Group XX (Name 1, Name 2, Name 3, Name 4)

# Homework 4

The focus of this homework is projections and least square regression.
Let's start with intialization as usual.

In [1]:
import numpy as np                # basic arrays, vectors, matrices
import scipy as sp                # matrix linear algebra 

import matplotlib                 # plotting
import matplotlib.pyplot as plt   # plotting
from mpl_toolkits.mplot3d import Axes3D
%matplotlib inline
from IPython.core.display import HTML
HTML("""<style>.output_png { display: table-cell; text-align: center; vertical-align: middle; }</style>""");

- - -

<div class="alert alert-info">

### Projections in 3D
</div>

In the lecture, we discussed how to project vectors onto subspaces. In this exercise we try the idea by projecting a 3D parametric curve on a plane. In general, parameteric curves are given as 

\begin{eqnarray}
x_1 & = & x_1(t) \\
x_2 & = & x_2(t) \\
x_3 & = & x_3(t)
\end{eqnarray}

As an example, the equation of a helix on the cylinder $x_2^2 + x_3^2 = 1$ is given by
\begin{eqnarray}
x_1 & = & t \\
x_2 & = & \cos(2t) \\
x_3 & = & \sin(2t)
\end{eqnarray}

Below, we define such a helix and plot it

In [ ]:
# we define the coordinates of a helix 
N = 1000
t = np.linspace(0, 12, N)
x1 = t
x2 = np.cos(2*t)
x3 = np.sin(2*t)
plt.figure(figsize=(10,8))
ax = plt.axes(projection='3d')
ax.scatter(x1, x2, x3)


Using parameterized curves, one can generate more fancy shapes: 

In [ ]:
N = 2456
t = np.linspace(0, 12, N)
x1 = 3/2*t + abs(t-1) - abs(t-3) + abs(t-4) + abs(t-7) - 2*abs(t-8) + 1.5* abs(t-11) \
   + 0.5 * (abs(t-4)/(t-4)  - 3*abs(t-6)/(t-6) + abs(t-7)/(t-7)  - 2*abs(t-8)/(t-8) ) \
   + 0.5 * (3*abs(t-9)/(t-9) +  3*abs(t-10)/(t-10) - 3*abs(t-11)/(t-11)   )   - 37/2


x2 = 2*t - 3*abs(t-1) + 2*abs(t-2) - 3*abs(t-3) + 4*abs(t-4) - 4*abs(t-5) + 2*abs(t-6) - 2*abs(t-8)+ 2*abs(t-11) \
   + abs(t-6)/(t-6)+ abs(t-7)/(t-7) + 2*abs(t-9)/(t-9) + 2*abs(t-10)/(t-10) +abs(t-11)/(t-11)-1

x3 = N*[1.0]
plt.figure(figsize=(10,8))
ax = plt.axes(projection='3d')
ax.scatter(x1, x2, x3)

On the other hand, the coordinate of a point on a plane in 3D coordinate system is given by
\begin{equation}
\mathbf{x} = (x_1, x_2, x_3)^\top = \mathbf{r_0} + \alpha \mathbf{v_1} + \beta \mathbf{v_2},
\end{equation}
where $\mathbf{r_0},\mathbf{v_1},\mathbf{v_2}$ are vectors and $\alpha, \beta$ are real numbers. Note that the vectors $\mathbf{v_1},\mathbf{v_2}$ span a subspace (a plane), which goes through the origin. The vector $\mathbf{r_0}$ shifts that plane in the coordinate sytem.  

<div class="alert alert-success">

**Task**: Complete the function `project_points_on_a_surface` below to implement the projection on a plane, given the coordinate matrix `C`, which holds the three coordinates of each point in its three rows. Note that each point is a column of this matrix. The function should have the following properties:
- The function should check whether v1 and v2 are linearly independent. If this is not the case it should terminate with error.
- Within the function body, a projection matrix P should be calculated (using the input arguments of the function). This matrix is then used to perform the projection.
- The function should return the coordinates of the projected points as a matrix. The input paramaters of the function are the coordinate matrix of input points, the vectors r0,v1 and v2 that are used to parameterize the projection plane. 
- Note: You may use standard functions like the numpy.linalg.inv
</div>

In [ ]:
# we fill the matrix C with coordinate arrays 
C = np.array([x1,x2,x3])

def project_points_on_a_surface(C, r0, v1, v2):
    """project the curve onto the surface parameterized by v1 and v2"""    

    # TODO
    
    return Cprojected


r0 = np.array([0,0,1])
v1 = np.array([1,2,3])
v2 = np.array([1,1,1])


Cprojected = project_points_on_a_surface(C,r0,v1,v2)

# Now we plot the projected coordinates along with the original coordinates.
# Optional task: Install the PyQt5 module, uncomment the following line and 
# add proj_type="ortho" to the plt.axes call, to obtain an interactive 3D plot.
# This allows you to visually check whether your solution is correct.
# %matplotlib qt 
plt.figure(figsize=(10,8))
ax = plt.axes(projection='3d') # or plt.axes(projection='3d',proj_type="ortho") 
ax.scatter(C[0,:], C[1,:], C[2,:])
ax.scatter(Cprojected[0,:], Cprojected[1,:], Cprojected[2,:])



<div class="alert alert-info">
    
### Least Squares Regression 
</div>

As discussed in the course, the least squares method can be used to solve overdetermined systems (these have more equations than unknowns). The method has usage in many applications, especially in statistics and machine learning. In this exercise, we try it in the context of machine learning (ML).

In this task, the goal is to develop a predictive ML model based on the so-called Boston Housing data. The dataset can be found in the UCI Machine Learning Repository. The data was collected in 1970s and each entry (each row in the data) represents information about different features of homes from various suburbs located in Boston. Let's load the data and take a first look.

<div class="alert alert-danger">
    
**Note:** Please download the file `Boston_Housing_Data.csv` from the OLAT materials folder, and put it in the same directory as this notebook.
</div>

In [2]:
from numpy import genfromtxt
housingData = genfromtxt('Boston_Housing_Data.csv', delimiter=',')
NumberOfSamples = housingData.shape[0]
NumberOfFeatures = housingData.shape[1]
print('Boston Housing Data:',housingData)
print('The data has', NumberOfSamples, 'samples.')
print('Each sample has', NumberOfFeatures, 'features')


Boston Housing Data: [[6.3200e-03 1.8000e+01 2.3100e+00 ... 3.9690e+02 4.9800e+00 2.4000e+01]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 3.9690e+02 9.1400e+00 2.1600e+01]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 3.9283e+02 4.0300e+00 3.4700e+01]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 3.9690e+02 5.6400e+00 2.3900e+01]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 3.9345e+02 6.4800e+00 2.2000e+01]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 3.9690e+02 7.8800e+00 1.1900e+01]]
The data has 506 samples.
Each sample has 14 features


After loading, we can observe that the `housingData` array contains 506 samples (rows). Each sample represents a house and has 14 features (columns in the data). The features are like:
- capita crime rate 
- index of accessibility to radial highways
- pupil-teacher ratio by town
- ...

and so on. 

What is most important to us is the last feature (last column in the data), namely the price of the house in thousand dollars (do not surprise much, these are 70s prices!). Our aim is to develop a model, which can predict the value of any house. For this exercise, we choose the linear model for the price denoted by $p$:

\begin{equation}
p(\mathbf{x}) = w_0 + w_1 * x_1 + w_2 * x_2 + \ldots + w_{13}* x_{13}, 
\end{equation}

where the vector $\mathbf{w} = (w_0,w_1,\ldots, w_{13})^\top$ are the weights of the model and the vector $\mathbf{x} = (x_1,x_2, \ldots, x_{13})^\top$ is the features vector. To account for the bias term $w_0$, we can add $x_0=1$ to the features vector such that the price function is simply a scalar product:

\begin{equation}
p(\mathbf{x}) = w_0 * x_0 + w_1 * x_1 + w_2 * x_2 + \ldots + w_{13}* x_{13} = \mathbf{w}^\top \mathbf{x}, 
\end{equation}

Our price model can predict the value of any house if its other $13$ features are known. The crucial issue is now to adjust the model weights in such way that our model performs well for the new samples, which are not in the data. To find the best values for the weights, we take $N$ sample points from the data (these we call training samples in the following) and try to fit the linear price function such that it provides exact interpolations at those points. Therefore, for each sample we can write a linear equation:

\begin{eqnarray}
w_0 * x_0^{(1)} + w_1 * x_1^{(1)} + w_2 * x_2^{(1)} + \ldots + w_{13}* x_{13}^{(1)} & = & p(\mathbf{x}^{(1)}) \\
w_0 * x_0^{(2)} + w_1 * x_1^{(2)} + w_2 * x_2^{(2)} + \ldots + w_{13}* x_{13}^{(2)} & = & p(\mathbf{x}^{(2)}) \\
\vdots \quad \quad \quad \quad \vdots & & \vdots \\
w_0 * x_0^{(N)} + w_1 * x_1^{(N)} + w_2 * x_2^{(N)} + \ldots + w_{13}* x_{13}^{(N)} & = & p(\mathbf{x}^{(N)})
\end{eqnarray}

The superscript $(i)$ denotes the index of the particular sample, e.g., $x_3^{(11)}$ means the $x_3$ from the $11$th sample. In matrix form we can write the above linear system as

\begin{equation}
X \mathbf{w} = \mathbf{p},
\end{equation}

with the coefficient matrix

\begin{equation}
X=
\begin{bmatrix}
1 & x_1^{(1)} & x_2^{(1)} & \ldots & x_{13}^{(1)} \\
1 & x_1^{(2)} & x_2^{(2)} & \ldots & x_{13}^{(2)} \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
1 & x_1^{(N)} & x_2^{(N)} & \ldots & x_{13}^{(N)}
\end{bmatrix}.
\end{equation}

Note that for $N>14$ this linear system is overdetermined and it is very unlikely that we can find a solution in the classical sense. However, as we have learnt in the lecture, overdetermined systems can be solved using the least squares method. Finding this least square solution is precisely the task in this exercise.

As you might have noted, we have in total $506$ samples in the data. We use $N=400$ samples for the linear system. The rest $106$ samples will be used to evaluate the performance our model.

In [3]:
N = 400
trainingData = housingData[0:N,:]
testData = housingData[N:NumberOfSamples,:]
print(trainingData)

[[6.32000e-03 1.80000e+01 2.31000e+00 ... 3.96900e+02 4.98000e+00
  2.40000e+01]
 [2.73100e-02 0.00000e+00 7.07000e+00 ... 3.96900e+02 9.14000e+00
  2.16000e+01]
 [2.72900e-02 0.00000e+00 7.07000e+00 ... 3.92830e+02 4.03000e+00
  3.47000e+01]
 ...
 [7.67202e+00 0.00000e+00 1.81000e+01 ... 3.93100e+02 1.99200e+01
  8.50000e+00]
 [3.83518e+01 0.00000e+00 1.81000e+01 ... 3.96900e+02 3.05900e+01
  5.00000e+00]
 [9.91655e+00 0.00000e+00 1.81000e+01 ... 3.38160e+02 2.99700e+01
  6.30000e+00]]


<div class="alert alert-success">

**Task**: Implement the functions `find_weights` and `estimate_price` below
- The function find_weights should find the least square solution for the problem $X\mathbf{w} = \mathbf{p}$
- The function `estimate_price` should simply return a price value for given $\mathbf{w}$ and $\mathbf{x}$.
- Note: You may use standard functions like the numpy.linalg.inv but do not use numpy.linalg.lstsq
- Note: Do not forget the column of ones in the $X$ matrix (for the bias term)
</div>

In [4]:
def find_weights(trainingData):
   
    # TO DO
    p = trainingData[:, [-1]]
    X = np.copy(trainingData)
    X[:,-1] = 1
    X_transpose = X.transpose()
    final_X     = np.matmul(X_transpose, X)
    inverse_X   = np.linalg.inv(final_X)
    X_new       = np.matmul(inverse_X, X_transpose)
    weights     = np.matmul(X_new, p) 

    
    return weights


def estimate_price(X , weights):
    
    # TO DO
    price = np.dot(X , weights )
    
    return price

'''print(price)
print(price.shape)
print(weights)
print(weights.shape)'''


'print(price)\nprint(price.shape)\nprint(weights)\nprint(weights.shape)'

Now we are ready to try our model with the test data. First we calculate the weight vector by calling the function 
`find_weights` with the `trainingData`:

In [5]:
weights = find_weights(trainingData)

In the next step, we try our model on the `testData` and calculate the mean squared error (MSE) of our estimations:

In [6]:
MSE = 0.0
for sample in testData:
    x = sample[0:NumberOfFeatures-1]
    x = np.append([1.0],x)
    x = np.expand_dims(x, axis=0) # hatte mit den Dimensionen nicht gepasst. x war (14,) 
    #und estimatedPrice war (14,1) deshalb hab ich die Zeile noch hinzugefügt 
    '''
    print(x)
    print (x.shape)
    print(weights)
    print(weights.shape)'''
    estimatedPrice = estimate_price(weights,x)
    truePrice = sample[NumberOfFeatures-1]  
    print('estimated price = ',estimatedPrice, 'true price = ',truePrice)
    MSE = MSE + (estimatedPrice-truePrice)**2.0 

MSE = MSE/(NumberOfSamples-N)
print('Mean Squared Error for the test data = ',MSE)

estimated price =  [[-1.91246374e-01 -4.78997581e+00  0.00000000e+00 -3.46155937e+00
   0.00000000e+00 -1.32533737e-01 -1.14499204e+00 -1.91246374e+01
  -3.03852239e-01 -4.58991298e+00 -1.27370085e+02 -3.86317676e+00
  -7.59056859e+01 -5.11966543e+00]
 [ 4.42289967e-02  1.10776388e+00  0.00000000e+00  8.00544841e-01
   0.00000000e+00  3.06506947e-02  2.64799003e-01  4.42289967e+00
   7.02710300e-02  1.06149592e+00  2.94565118e+01  8.93425734e-01
   1.75544888e+01  1.18401024e+00]
 [ 5.52207977e-02  1.38306562e+00  0.00000000e+00  9.99496439e-01
   0.00000000e+00  3.82680128e-02  3.30606916e-01  5.52207977e+00
   8.77348034e-02  1.32529915e+00  3.67770513e+01  1.11546011e+00
   2.19171346e+01  1.47826075e+00]
 [ 1.71631351e+00  4.29869598e+01  0.00000000e+00  3.10652745e+01
   0.00000000e+00  1.18940526e+00  1.02755690e+01  1.71631351e+02
   2.72687890e+00  4.11915242e+01  1.14306480e+03  3.46695329e+01
   6.81204832e+02  4.59457127e+01]
 [-1.49957220e+01 -3.75584352e+02  0.00000000e+00